### ISSUE TITLE
investigate changes to k8s plugin 0.12 and use of template namespaces

### ISSUE THREAD

**Issue by gabemontero** [openshift/jenkins#342]
found by @scoheb, see https://github.com/jenkinsci/kubernetes-plugin/pull/179

**Comment by gabemontero** [comment]
Possibly can work around with the sync plugins pod template support

**Comment by scoheb** [comment]
via `ConfigMapWatcher.podTemplatesFromConfigMap()` ?

**Comment by gabemontero** [comment]
On Tue, Aug 1, 2017 at 10:21 AM, Scott Hebert <notifications@github.com>
wrote:

> via ConfigMapWatcher.podTemplatesFromConfigMap() ?
>
Yep, along with the image stream watcher we have.

**Comment by scoheb** [comment]
I think it will just work by changing to use 0.12 of the kubernetes plugin...namespace will be pre-filled when you receive the configmap and xstream should just handle it.

**Comment by gabemontero** [comment]
actually I'm going to look at bumping all the way to 1.0 today following some regression testing

but yeah, per your https://github.com/openshift/jenkins/issues/342#issuecomment-319387075 we should be good

i'll close out with the PR that bring 1.0 in

**Comment by gabemontero** [comment]
we'll that blew up fast:

A bunch of exceptions like:
```
imer task io.fabric8.jenkins.openshiftsync.BuildConfigWatcher$1@926ab9 failed
java.lang.NoClassDefFoundError: Could not initialize class io.fabric8.openshift.client.dsl.internal.BuildConfigOperationsImpl
	at io.fabric8.openshift.client.DefaultOpenShiftClient.buildConfigs(DefaultOpenShiftClient.java:317)
	at io.fabric8.jenkins.openshiftsync.BuildConfigWatcher$1.doRun(BuildConfigWatcher.java:77)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:51)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Will go back to 0.10, get a base line, then try something between 0.12 and 1.0

**Comment by gabemontero** [comment]
Actually there is nothing between 0.12 and 1.0.

And 0.12 of the kubernetes plugin has similar class loading issues as 1.0:

```
Sep 20, 2017 4:31:48 PM SEVERE hudson.triggers.SafeTimerTask run
Timer task io.fabric8.jenkins.openshiftsync.BuildWatcher$1@d5b75d failed
java.lang.NoClassDefFoundError: Could not initialize class io.fabric8.openshift.client.dsl.internal.OpenShiftOperation
	at io.fabric8.openshift.client.DefaultOpenShiftClient.builds(DefaultOpenShiftClient.java:312)
	at io.fabric8.jenkins.openshiftsync.BuildWatcher$1.doRun(BuildWatcher.java:76)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:51)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Sep 20, 2017 4:31:48 PM SEVERE hudson.triggers.SafeTimerTask run
Timer task io.fabric8.jenkins.openshiftsync.ConfigMapWatcher$1@14c5b8d failed
java.lang.LinkageError: loader constraint violation: when resolving method "com.fasterxml.jackson.databind.ObjectMapper.<init>(Lcom/fasterxml/jackson/core/JsonFactory;)V" the class loader (instance of hudson/PluginFirstClassLoader) of the current class, io/fabric8/kubernetes/client/utils/Serialization, and the class loader (instance of hudson/PluginFirstClassLoader) for the method's defining class, com/fasterxml/jackson/databind/ObjectMapper, have different Class objects for the type com/fasterxml/jackson/core/JsonFactory used in the signature
	at io.fabric8.kubernetes.client.utils.Serialization.<clinit>(Serialization.java:38)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.<clinit>(OperationSupport.java:59)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.configMaps(DefaultKubernetesClient.java:254)
	at io.fabric8.openshift.client.DefaultOpenShiftClient.configMaps(DefaultOpenShiftClient.java:302)
	at io.fabric8.jenkins.openshiftsync.ConfigMapWatcher$1.doRun(ConfigMapWatcher.java:66)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:51)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

both the openshift sync and k8s plugins have the `<pluginFirstClassLoader>true</pluginFirstClassLoader>` so at first blush not sure why the fabric versions are out of whack when I upgrade k8s beyon 0.10

@jstrachan - any chance you or someone on your team is testing with the later fabric versions and k8s 0.12 or 1.0 (and might have seen something like this) ?

Pending that, or if you have some time to dig @scoheb , I'm going to have to put bumping k8s beyond 0.10 for a bit.  Need to circle back to higher priority items.

@bparees fyi

**Comment by gabemontero** [comment]
ok, the reverting of sync plugin's fabric8 version back to 2.3.1 (to work around an openshift bug) has resolved the class loading related issues noted last week;

also, v1.0 of k8s plugin has @scoheb 's fix; v0.12 does not

given the successful regression testing I am seeing, will be submitting a pull for the jenkins image to bump to 1.0
